artifact/v3/header/header_test.cpp

```diff
@@ -446,3 +448,176 @@ TEST_F(HeaderTestEnv, TestHeaderFilesOutOfOrder) {
 		expected_header.error().message,
 		"Got unexpected token: 'type-info' expected 'header-info'");
 }
+
+TEST_F(HeaderTestEnv, TestHeaderMetaDataSuccess) {
+	stringstream meta_data {
+		R"(
+{
+  "foo": "bar",
+  "bar": "100",
+  "baz": 1,
+  "bur": ["foo", 1000]
+}
+)"};
+
+	mender::common::io::StreamReader sr {meta_data};
+
+	auto expected_meta_data = header::meta_data::Parse(sr);
+
+	ASSERT_TRUE(expected_meta_data) << expected_meta_data.error().message;
+}
+
+TEST_F(HeaderTestEnv, TestHeaderMetaDataParsingTopLevelKeys) {
+	// Invalid, can only contain top-level keys
+	stringstream meta_data {
+		R"(
+["foo", "bar" ]
+)"};
+
+	mender::common::io::StreamReader sr {meta_data};
+
+	auto expected_meta_data = header::meta_data::Parse(sr);
+
+	ASSERT_FALSE(expected_meta_data);
+	EXPECT_EQ(expected_meta_data.error().message, "The meta-data needs to be a top-level object");
+}
+
+TEST_F(HeaderTestEnv, TestHeaderMetaDataParsingIntsStringsAndLists) {
+	// Invalid, can only contain strings, ints and lists of ints and strings
+	stringstream meta_data {
+		R"(
+{
+  "foo": { "bar": "baz" }
+}
+)"};
+
+	mender::common::io::StreamReader sr {meta_data};
+
+	auto expected_meta_data = header::meta_data::Parse(sr);
+
+	ASSERT_FALSE(expected_meta_data);
+	EXPECT_EQ(
+		expected_meta_data.error().message,
+		"The meta-data needs to only be strings, ints and arrays of ints and strings");
+}
+
+TEST_F(HeaderTestEnv, TestHeaderMetaDataIsMinified) {
+	// Invalid, can only contain strings, ints and lists of ints and strings
+	stringstream meta_data {
+		R"(
+{
+  "foo": { "bar": "baz" }
+}
+)"};
+
+	mender::common::io::StreamReader sr {meta_data};
+
+	auto expected_meta_data = header::meta_data::Parse(sr);
+
+	ASSERT_FALSE(expected_meta_data);
+	EXPECT_EQ(
+		expected_meta_data.error().message,
+		"The meta-data needs to only be strings, ints and arrays of ints and strings");
+}
+
+// TEST_F(HeaderTestEnv, TestHeaderMetaDataIsMinifiedFailure) {
```

Comment:
Still a little confused, should we verify that it is minified or not?

Reply:
I suspect we don't need to. It's mender-artifact's responsibility to make sure that it is minified. If the artifact happens to be handcrafted I suppose it might not be minified, but I don't think the client should care.

When we put the meta-data in the Update Module file structure, it needs to be line separated though. This is important due to the "dumb" parsing done by many shell scripts.